### ISSUE TITLE
Turning orientationOnByDefault with javascript

### ISSUE THREAD

**Issue by DStillingfleet** [mpetroff/pannellum#478]
Hi Matthew

Thank you for your great work.

I'm using API to load an pannellum instance, and will set the orientationOnByDefault to false so that I can manipulate the panorama with lookAt, setYaw, etc.  When these actions are complete, I wish to set the orientationOnByDefault to true so that the panorama will pan / tilt with the orientation of the mobile device.  

Can you provide some guidance please.

Kind regards

Derek

**Comment by DStillingfleet** [comment]
P.S.

Is there a way to set the direction of rotation with lookAt?

**Comment by mpetroff** [comment]
This was literally just asked for in #477, so I added `isOrientationSupported()`, `startOrientation()`, and `stopOrientation()` functions to the API in 35edd9f5f7d7900515ee77d155c7d9a7683203c8.

The `lookAt` rotates in the direction that's shortest; this is not configurable (there's a bug in 2.3.2 for the rotation direction, that was fixed in 129bc0310d71a1d483c4bdbcf01c266fde98d767).

**Comment by DStillingfleet** [comment]
Thank you Matthew.

This will be the first time I have tried to downloads a development version from GitHub.  

I've copied across pannellum.css, RequestAnimationFrame.js, libpannellum.js, pannellum.js, and pannellum.htm from the src directories, but when I try my webpage pannellum is stuck in loading the panorama.  What am I missing?

The bug fix for the shortest direction of rotation in LookAt is exactly what I was hoping to achieve.

Best wishes


Derek

**Comment by mpetroff** [comment]
What does your browser's developer console say?

**Comment by DStillingfleet** [comment]
The developer console reports...

ReferenceError: libpannellum is not defined      Pannellum.js:430:9

**Comment by DanielBiegler** [comment]
Are you loading the scripts in the right order? Like so:

```
.
.
<script type="text/javascript" src="../js/libpannellum.js"></script>
<script type="text/javascript" src="../js/RequestAnimationFrame.js"></script>
<script type="text/javascript" src="../js/pannellum.js"></script>
<script type="text/javascript" src="standalone.js"></script>
.
.
```

**Comment by DStillingfleet** [comment]
Ah, I didn't realise I had to load the scripts myself. Had assumed that pannellum.htm did that.  I'll give it a try.

Thank you.

**Comment by mpetroff** [comment]
The API isn't designed to be used with `pannellum.htm`. If you're using the API, you should be loading the JS and CSS files directly and not using `pannellum.htm`.

**Comment by DStillingfleet** [comment]
Many thanks,  I followed the advice and add the panorama is loading fine and functioning well.  

I hadn't realised that the pannellum.htm was not being used with the API, but now you said it, it makes sense.

Maybe I misunderstood, but lookAt function is not always rotating through the shortest angle as suggested above.

**Comment by mpetroff** [comment]
I thought I had fixed that, but apparently I didn't. It should work now as of 9c5d3212a63fa79bb5e3397aad114ba602252948.

**Comment by DStillingfleet** [comment]
Yes, it's fixed now, thank you so much.

**Comment by DStillingfleet** [comment]
I had not tried this previously with this version of pannellum.js, but as soon as one turns on device orientation, either via the icon, or via the new API command, the panorama becomes horizontal bands of colour curving as if within a sphere. Any idea why?

**Comment by mpetroff** [comment]
There have been other issues where people have reported horizontal bands, but it's always been immediate, not a side effect of turning on device orientation control. Does this behavior persist if device orientation control is turned back off? What browser / device are you using?

**Comment by DStillingfleet** [comment]
When device orientation is turned off, the image is restored.

I'm using an iPhone 6 to test the code, both chrome and safari show the same behaviour.

Regards


Derek

**Comment by DStillingfleet** [comment]
I don't know if it is relevant, but the web console is reporting "TypeError: document.getElementById(...) is null" in standalone.js 5:5.

**Comment by mpetroff** [comment]
I'm confused. I thought you were using the API, not the standalone version. As the name implies, `standalone.js` is only used with `pannellum.htm` and the standalone version.

Since Chrome on iOS is simply a different skin around Safari, the same behavior between the two browsers is expected. What version of iOS are you running?

**Comment by DStillingfleet** [comment]
When I had problems first running the developer version of the script, I was advised to make sure I loaded the scripts in order, and standalone.js was included on the list the scripts should be loaded in.I've removed the call to include the standalone.js, but the banding problem still occurs.Here's a link to the webpage if it helps...http://www.virtualmountains.co.uk/LakeDistrictV20/TourControlV20-300.html.On 24 September 2017 at 22:50 Matthew Petroff <notifications@github.com> wrote:I'm confused. I thought you were using the API, not the standalone version. As the name implies, standalone.js is only used with pannellum.htm and the standalone version.Since Chrome on iOS is simply a different skin around Safari, the same behavior between the two browsers is expected. What version of iOS are you running?--You are receiving this because you authored the thread.Reply to this email directly or view it on GitHub:https://github.com/mpetroff/pannellum/issues/478#issuecomment-331742718

**Comment by mpetroff** [comment]
After seeing the error, it was obvious to me that the yaw was being set to an invalid value (iOS showed horizontal band, while Chrome on Android just went black).

It took a bit to track down the cause: your configuration has `northOffset` set as a string; changing it to a number (dropping the quotation marks) should fix this.

Internally, Pannellum adds other numbers to the `northOffset` value. Since JavaScript does implicit type conversion and uses `+` for string concatenation, when a number is added to a string, the number is cast to a string and concatenated with the first string; when later code tries to perform trig with this resulting string, JavaScript attempts to cast back to a number. For positive numbers this provides an incorrect result but does not cause an error, e.g. `"298" + 23.3` becomes `"29823.3"`, which is then cast back to a number (although incorrect). However, if the number is negative, the result can't be cast back to a number, e.g. `"298" + (-23.3)` becomes `"298-23.3"`; this becomes a `NaN`, causing the rendering artifact.

**Comment by DStillingfleet** [comment]
Matthew, that is fantastic, I think you've gone beyond the call of duty to help me by looking into my code, it's unexpected and very much appreciated.

I hardly dare push my luck and ask another question, but...

What is the expected action of turning on the orientationOnByDefault?

My reason for asking is that I was expecting the yaw to remain set at the same value as when the orientation was turned on, and for the panorama to pan left or right as the orientation changed. I was wrong to do so.

Instead the yaw angle often jumps to a different angle, the rules for which I've tried to figure out but cannot, even after setting the North offset to zero.

**Comment by mpetroff** [comment]
I'm not quite sure what you mean. The `orientationOnByDefault` parameter isn't something that's turned on or off, per se. It's part of the initial configuration; when `true`, it automatically turns device orientation control on when the panorama loads. Do you mean the `startOrientation()` method I just added?

**Comment by DStillingfleet** [comment]
Dirty, yes, I mean the startOrientation() method just added.

**Comment by mpetroff** [comment]
The behavior you expected was the intended behavior. However, it seems that Safari (iOS) and Chrome (Android) implement device orientation events differently; Safari resets the yaw angle each time the listener is removed and readded, while Chrome maintains the same reference frame throughout the session. The existing code only worked in one case. I think I've fixed it in f0fcf31998aa149212546dd07276bd93bc01a01c.

**Comment by DStillingfleet** [comment]
Thank you Matthew, it's fixed.

All the best

Derek
